APMrover2/mode_breadcrumb.cpp

```diff
@@ -0,0 +1,183 @@
+#include "mode.h"
+#include "Rover.h"
+
+
+bool ModeBreadcrumb::_enter()
+{
+    //don't call enter again if _crumbs exists
+    if (_crumbs != nullptr) {
+        return false;
+    }
+
+    if (_crumbs_sem == nullptr){
+        // create semaphore
+        _crumbs_sem = hal.util->new_semaphore();
+        if (_crumbs_sem == nullptr) {
+            return false;
+        }
+    }
+
+    _current_crumb = nullptr;
+
+    // allocate crumbs queue
+    _crumbs = (Location*)calloc(crumbs_max, sizeof(Location));
+
+    _crumbs_tail = 0;
+    _crumbs_head = 0;
+
+    lonely_mode = nullptr;
+
+    //reuse guided mode
+    if (!ModeGuided::_enter()) {
+        free(_crumbs);
+        return false;
+    }
+
+    return true;
+}
+
+void ModeBreadcrumb::_exit()
+{
+    //reset lonely mode in case we go back into breadcrumb mode
+    lonely_mode = nullptr;
+
+    _current_crumb = nullptr;
+
+    if (_crumbs != nullptr) {
+        free(_crumbs);
```

Comment:
We may want to revisit the dynamic allocation model here.  Allocating at runtime can do bad things to your vehicle; you can start to run out of memory unexpectedly.  The only "fix" for that is to allow the user to specify a number of crumbs as a parameter.

I wouldn't stop this PR going in because of this.